### PR TITLE
Publisher Collection Rework

### PR DESCRIPTION
**Pull request by addie9800**
This PR replaces the `PublisherEnum`, `PublisherCollection` and `PublisherSpec` types with two new types: `Publisher` and `PublisherGroup`. 

Note: I haven't done the documentation yet, I thought it might be better to first get the code approved and then update the documentation to avoid unnecessary double work.

### CODE REVIEW

File: tests/utility.py [hunk 1]
```diff
@@ -268,27 +268,29 @@ def load_html_test_file_mapping(publisher: PublisherEnum) -> Dict[Type[BaseParse
...
 
-def get_meta_info_file(publisher: PublisherEnum) -> JSONFile[Dict[str, Dict[str, Any]]]:
+def get_meta_info_file(publisher: Publisher) -> JSONFile[Dict[str, Dict[str, Any]]]:
     return JSONFileWithExtractionDecoderEncoder(generate_meta_info_path(publisher))
 
 
-def generate_parser_test_case_json_path(publisher: PublisherEnum) -> Path:
-    return generate_absolute_section_path(publisher) / f"{publisher.name}.json"
+def generate_parser_test_case_json_path(publisher: Publisher) -> Path:
+    return generate_absolute_section_path(publisher) / f"{publisher.publisher_name}.json"
```
Comment (MaxDall) [inline tests/utility.py:286]: We might rewrite this part so that we don't need to keep `publisher_name`.

**Review by MaxDall — CHANGES_REQUESTED** [review]
Thanks for tackling one of the open problems ;) I would suggest going over this in iteration and for now I tried to address most of the structural/basic problems.

File: src/fundus/publishers/base_objects.py [hunk 2]
```diff
@@ -80,142 +78,143 @@ def supports(self, source_types: List[Type[URLSource]]) -> bool:
...
-    >>> class PoliticalPublisher(PublisherEnum):
-    >>>     ...
-    >>>
-    >>> class NewCollection(metaclass=PublisherCollectionMeta):
-    >>>     political = PoliticalPublisher
 
-    You can still use methods or non-PublisherEnum class attributes, e.g.
+class PublisherGroup(type):
+    _length: int
+    _contents: Set[str]
```
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py:84]: The idea was that those two elements are needed every once in a while and I thought this way we would avoid iterating through all subgroups to get the length and similarly for the direct attributes of the PublisherGruoup

File: src/fundus/publishers/base_objects.py [hunk 3]
```diff
@@ -80,142 +78,143 @@ def supports(self, source_types: List[Type[URLSource]]) -> bool:
...
+                return False
+        else:
+            search_string = __x
+        if search_string in self._contents:
+            return True
+        for element in self._contents:
+            if isinstance(attribute := getattr(self, element), PublisherGroup):
+                if search_string in attribute:
+                    return True
+        return False
```
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py:146]: Good Question, I am not sure we used it, but it was part of the previous implementation, so I decided to keep it :)

File: src/fundus/publishers/base_objects.py [hunk 4]
```diff
@@ -80,142 +78,143 @@ def supports(self, source_types: List[Type[URLSource]]) -> bool:
...
-            yield from enum
-
-    def __getitem__(self, name: str) -> PublisherEnum:
+        for element_name in self._contents:
+            if isinstance(element := getattr(self, element_name), Publisher):
+                yield element
+            elif isinstance(element, PublisherGroup):
+                yield from element
+            else:
+                raise AttributeError(f"Element {element} of invalid type {type(element)}")
```
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py:161]: Sounds good

File: docs/supported_publishers.md [hunk 5]
```diff
@@ -16,10 +16,10 @@
   <tbody>
     <tr>
       <td>
-        <code>DerStandard</code>
+        <code>Der Standard</code>
```
[the remark this answers is not in the document]
Comment (addie9800) [reply docs/supported_publishers.md:19]: Yep, they did, it's fixed

File: scripts/publisher_coverage.py [hunk 6]
```diff
@@ -19,18 +18,16 @@
...
 
     for publisher_region in publisher_regions:
         print(f"{publisher_region.__name__:-^50}")
 
-        publisher: PublisherEnum
-        for publisher in sorted(
-            publisher_region, key=lambda p: cast(PublisherEnum, p).name  # type: ignore[no-any-return]
-        ):
-            publisher_name: str = publisher.name  # type: ignore[attr-defined]
+        publisher: Publisher
```
[the remark this answers is not in the document]
Comment (addie9800) [reply scripts/publisher_coverage.py:28]: Tbh, I don't think there are any left that are related to the publisher enum

File: src/fundus/publishers/base_objects.py [hunk 7]
```diff
@@ -1,53 +1,48 @@
...
+        """
+        Initialization of a new Publisher object
+
+        @param name: Name of the publisher, as it would appear on the website
+        @param domain: The domain of the publishers website
+        @param parser: Corresponding ParserProxy Object
+        @param sources: List of sources for articles from the publishers
+        @param query_parameter: Dictionary of query parameter: content to be appended to crawled URLs
+        @param url_filter: Regex filter to apply determining URLs to be skipped
+        @param request_header: Request header to be used for the GET-request
```
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py:31]: Ah yeah. I think I knew that at some point... I have fixed it.

File: src/fundus/publishers/base_objects.py [hunk 8]
```diff
@@ -1,53 +1,48 @@
...
+        self.parser = parser()
+        self.domain = domain
+        self.sources = sources
+        self.query_parameter = query_parameter
+        self.url_filter = url_filter
+        self.request_header = request_header
+        self.contained_in: Optional[str] = None
+        # This variable has been chosed for backwards compatibility, where publisher_name used to be the name of the
+        # publisher in the PublisherEnum
+        self.publisher_name: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py]: But don't we need both names? At the very least in the supported_publishers.md file both names are referenced. I fixed the rest :)

**Comment by addie9800** [comment]
Thanks for the very thorough review :)

File: src/fundus/publishers/base_objects.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (addie9800) [reply src/fundus/publishers/base_objects.py]: I modified the utility function slightly, this way we don't need the extra attribute at all. But I guess it's also not the most elegant solution

**Comment by addie9800** [comment]
I took the chance to also integrate the suggested changes, closing #498

**Comment by MaxDall** [comment]
@addie9800 I finished working on this, could you give it a look?

**Comment by addie9800** [comment]
Looks good as far as I can see. I updated the documentation to reflect the new old `_PublisherCollectionMeta`.

**Review by MaxDall — APPROVED** [review]
🚀

**Pull request merged**